lib/Provider/Command/MessageSend.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+namespace OCA\Mail\Provider\Command;
+
+use OCA\Mail\AppInfo\Application;
+use OCA\Mail\Db\LocalMessage;
+use OCA\Mail\Service\AccountService;
+use OCA\Mail\Service\Attachment\AttachmentService;
+use OCA\Mail\Service\OutboxService;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\IConfig;
+use OCP\Mail\Provider\Exception\SendException;
+use OCP\Mail\Provider\IAddress;
+use OCP\Mail\Provider\IMessage;
+
+class MessageSend {
+
+	public function __construct(
+		protected IConfig $config,
+		protected ITimeFactory $time,
+		protected AccountService $accountService,
+		protected OutboxService $outboxService,
+		protected AttachmentService $attachmentService
+	) {
+	}
+
+	/**
+	 * Performs send operation
+	 *
+	 * @since 4.0.0
+	 *
+	 * @param string $userId system user id
+	 * @param string $serviceId mail account id
+	 * @param IMessage $message mail message object with all required parameters to send a message
+	 * @param array $options array of options reserved for future use
+	 *
+	 * @return LocalMessage
+	 *
+	 * @throws SendException on failure, check message for reason
+	 *
+	 */
+	public function perform(string $userId, string $serviceId, IMessage $message, array $options = []): LocalMessage {
+		// find user mail account details
+		try {
+			$account = $this->accountService->find($userId, (int)$serviceId);
+		} catch (\Throwable $e) {
+			throw new SendException('Error: occurred while retrieving mail account details', 0, $e);
+		}
+		// convert mail provider message to mail app message
+		$localMessage = new LocalMessage();
+		$localMessage->setType($localMessage::TYPE_OUTGOING);
+		$localMessage->setAccountId($account->getId());
+		$localMessage->setSubject($message->getSubject());
+		$localMessage->setBody($message->getBody());
+		// disabled due to issues caused by opening these messages in gui
+		//$localMessage->setEditorBody($message->getBody());
+		$localMessage->setHtml(true);
+		$localMessage->setSendAt($this->time->getTime());
+		// convert all mail provider attachments to local attachments
+		$attachments = [];
+		if (count($message->getAttachments()) > 0) {
+			// iterate attachments and save them
+			foreach ($message->getAttachments() as $entry) {
+				// determine if required parameters are set
+				if (empty($entry->getName()) || empty($entry->getType()) || empty($entry->getContents())) {
```

Comment:
Thanks to psalm, I've looked closer at the given code.

If you patch the `IMipPlugin`  like below to pass a valid and an invalid attachment object, the first attachment is saved, the second one fails, but the first attachment is not removed.



```diff
Index: apps/dav/lib/CalDAV/Schedule/IMipPlugin.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/apps/dav/lib/CalDAV/Schedule/IMipPlugin.php b/apps/dav/lib/CalDAV/Schedule/IMipPlugin.php
--- a/apps/dav/lib/CalDAV/Schedule/IMipPlugin.php	(revision 8365f7b8acbeeec1c6a1c773a76f97fd38f2ce35)
+++ b/apps/dav/lib/CalDAV/Schedule/IMipPlugin.php	(date 1725395792585)
@@ -285,12 +285,20 @@
 				$message->setSubject($template->renderSubject());
 				$message->setBodyPlain($template->renderText());
 				$message->setBodyHtml($template->renderHtml());
-				$message->setAttachments((new \OCP\Mail\Provider\Attachment(
-					$itip_msg,
-					'event.ics',
-					'text/calendar; method=' . $iTipMessage->method,
-					true
-				)));
+				$message->setAttachments(
+					new \OCP\Mail\Provider\Attachment(
+						$itip_msg,
+						'event.ics',
+						'text/calendar; method=' . $iTipMessage->method,
+						true
+					),
+					new \OCP\Mail\Provider\Attachment(
+						null,
+						null,
+						null,
+						true
+					)
+				);
 				// send message
 				$mailService->sendMessage($message);
 			} else {

```

Reply:
Possible solution:

```diff
Index: lib/Provider/Command/MessageSend.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/Provider/Command/MessageSend.php b/lib/Provider/Command/MessageSend.php
--- a/lib/Provider/Command/MessageSend.php	(revision 83d3a0879b33c5483a9a408035ac411a907e69ef)
+++ b/lib/Provider/Command/MessageSend.php	(date 1725395506632)
@@ -9,6 +9,7 @@
 namespace OCA\Mail\Provider\Command;
 
 use OCA\Mail\AppInfo\Application;
+use OCA\Mail\Db\LocalAttachment;
 use OCA\Mail\Db\LocalMessage;
 use OCA\Mail\Service\AccountService;
 use OCA\Mail\Service\Attachment\AttachmentService;
@@ -17,9 +18,12 @@
 use OCP\IConfig;
 use OCP\Mail\Provider\Exception\SendException;
 use OCP\Mail\Provider\IAddress;
+use OCP\Mail\Provider\IAttachment;
 use OCP\Mail\Provider\IMessage;
 
 class MessageSend {
+	/** @var LocalAttachment[] */
+	private array $localAttachments = [];
 
 	public function __construct(
 		protected IConfig $config,
@@ -63,27 +67,14 @@
 		$localMessage->setHtml(true);
 		$localMessage->setSendAt($this->time->getTime());
 		// convert all mail provider attachments to local attachments
-		$attachments = [];
-		if (count($message->getAttachments()) > 0) {
-			// iterate attachments and save them
-			foreach ($message->getAttachments() as $entry) {
-				// determine if required parameters are set
-				if (empty($entry->getName()) || empty($entry->getType()) || empty($entry->getContents())) {
-					throw new SendException('Invalid Attachment Parameter: MUST contain values for Name, Type and Contents');
-				}
-				// convert mail provider attachment to mail app attachment
-				try {
-					$attachments[] = $this->attachmentService->addFileFromString(
-						$userId,
-						$entry->getName(),
-						$entry->getType(),
-						$entry->getContents()
-					)->jsonSerialize();
-				} catch (\Throwable $e) {
-					throw new SendException('Error: occurred while saving mail message attachment', 0, $e);
-				}
-			}
+
+		try {
+			$this->convertAttachments($userId, $message->getAttachments());
+		} catch (SendException $e) {
+			$this->cleanupAttachments($userId);
+			throw $e;
 		}
+
 		// determine if required To address is set
 		if (empty($message->getTo()) || empty($message->getTo()[0]->getAddress())) {
 			throw new SendException('Invalid Message Parameter: MUST contain at least one TO address with a valid address');
@@ -100,7 +91,7 @@
 				$to,
 				$cc,
 				$bcc,
-				$attachments
+				array_map(static fn(LocalAttachment $attachment) => $attachment->jsonSerialize(), $this->localAttachments)
 			);
 		} catch (\Throwable $e) {
 			throw new SendException('Error: occurred while saving mail message', 0, $e);
@@ -134,4 +125,39 @@
 		}, $addresses);
 	}
 
+	/**
+	 * @param IAttachment[] $attachments
+	 *
+	 * @throws SendException
+	 */
+	private function convertAttachments(string $userId, array $attachments): void
+	{
+		foreach ($attachments as $attachment) {
+			$name = $attachment->getName();
+			$type = $attachment->getType();
+			$contents = $attachment->getContents();
+
+			if (empty($name) || empty($type) || empty($contents)) {
+				throw new SendException('Invalid Attachment Parameter: MUST contain values for Name, Type and Contents');
+			}
+
+			try {
+				$this->localAttachments[] = $this->attachmentService->addFileFromString(
+					$userId,
+					$name,
+					$type,
+					$contents
+				);
+			} catch (\Throwable $e) {
+				throw new SendException('Error: occurred while saving mail message attachment', 0, $e);
+			}
+		}
+	}
+
+	private function cleanupAttachments(string $userId): void {
+		foreach ($this->localAttachments as $localAttachment) {
+			$this->attachmentService->deleteAttachment($userId, $localAttachment->getId());
+		}
+	}
+
 }

```

Reply:
Logic error find you did.

Logic error address I did, checking all data at the start

Reply:
> Example:
> 
> ```
> foreach ($message->getAttachments() as $attachment) {
>     $name = $attachment->getName();
>     $type = $attachment->getType();
>     $contents = $attachment->getContents();
> 
>     if (empty($name) || empty($type) || empty($contents)) {
>         throw new SendException('Invalid Attachment Parameter: MUST contain values for Name, Type and Contents');
>     }
> 
>     try {
>         $attachments[] = $this->attachmentService->addFileFromString(
>             $userId,
>             $name,
>             $type,
>             $contents
>         )->jsonSerialize();
>     } catch (\Throwable $e) {
>         throw new SendException('Error: occurred while saving mail message attachment', 0, $e);
>     }
> }
> ```

This didn't work, Psalm still wasn't happy, casting was the correct fix.

Reply:
> This didn't work, Psalm still wasn't happy, casting was the correct fix.

I tried the version before suggesting it, and it did work.

https://github.com/nextcloud/mail/pull/9651/commits/d423edb7e0968434d23d56fc7d9277d20a300767 did not work because the important piece was missing, the check if the variable $name is not empty. 

``` 
$name = $attachment->getName(); # possible types for $name: string and null
if (empty($name) { ...} # possible type for $name: string
```

Casting does the same, it changes the detected to string.

Reply:
> Logic error address I did, checking all data at the start

I considered the approach and withdrew it because it wasn't addressing the problem properly. The problem is looping over the attachments, creating records in mail_local_messages/on disk, and then interrupting the process and leaving some stale data. That might be an exception from the input validation, but can also happen when addFileFromString throws. Input validation is fine, addFileFromString is still open.

Please pick up my suggestion from https://github.com/nextcloud/mail/pull/9651#discussion_r1742666066 to resolve the problem. That's the same approach we are using for https://github.com/nextcloud/mail/blob/77cdc3861c0d8e9fb337e85084164c325485a315/lib/Controller/MessageApiController.php. 

Moving the logic into an own method also helps with the readability of the method and makes it easier to understand it.